database/pgsql/layer.go

```diff
@@ -18,300 +18,349 @@ import (
-		return -1, handleError("queryPersistLayer", err)
+	detectedBySet := mapset.NewSet()
+	for _, d := range detectedBy {
+		detectedBySet.Add(d)
 	}
 
-	if !id.Valid {
-		panic("null layer.id violates database constraint")
+	for _, f := range features {
+		if !detectedBySet.Contains(f.By) {
+			return database.ErrInvalidParameters
+		}
 	}
 
-	return id.Int64, nil
-}
-
-// PersistLayer relates layer identified by hash with namespaces,
-// features and processors provided. If the layer, namespaces, features are not
-// in database, the function returns an error.
-func (tx *pgSession) PersistLayer(hash string, namespaces []database.Namespace, features []database.Feature, processedBy database.Processors) error {
-	if hash == "" {
-		return commonerr.NewBadRequestError("Empty layer hash is not allowed")
+	for _, n := range namespaces {
+		if !detectedBySet.Contains(n.By) {
+			return database.ErrInvalidParameters
+		}
 	}
 
+	return nil
+}
+
+// PersistLayer saves the content of a layer to the database.
+func (tx *pgSession) PersistLayer(hash string, features []database.LayerFeature, namespaces []database.LayerNamespace, detectedBy []database.Detector) error {
 	var (
-		err error
-		id  int64
+		err         error
+		id          int64
+		detectorIDs []int64
 	)
 
-	if id, err = tx.persistLayer(hash); err != nil {
+	if err = sanitizePersistLayerInput(hash, features, namespaces, detectedBy); err != nil {
 		return err
 	}
 
-	if err = tx.persistLayerNamespace(id, namespaces); err != nil {
+	if id, err = tx.soiLayer(hash); err != nil {
+		return err
+	}
+
+	if detectorIDs, err = tx.findDetectorIDs(detectedBy); err != nil {
+		if err == commonerr.ErrNotFound {
+			return database.ErrMissingEntities
+		}
+
 		return err
 	}
 
-	if err = tx.persistLayerFeatures(id, features); err != nil {
+	if err = tx.persistLayerDetectors(id, detectorIDs); err != nil {
 		return err
 	}
 
-	if err = tx.persistLayerDetectors(id, processedBy.Detectors); err != nil {
+	if err = tx.persistAllLayerFeatures(id, features); err != nil {
 		return err
 	}
 
-	if err = tx.persistLayerListers(id, processedBy.Listers); err != nil {
+	if err = tx.persistAllLayerNamespaces(id, namespaces); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (tx *pgSession) persistLayerDetectors(id int64, detectors []string) error {
-	if len(detectors) == 0 {
-		return nil
+func (tx *pgSession) persistAllLayerNamespaces(layerID int64, namespaces []database.LayerNamespace) error {
+	detectorMap, err := tx.findAllDetectors()
+	if err != nil {
+		return err
 	}
 
-	// Sorting is needed before inserting into database to prevent deadlock.
-	sort.Strings(detectors)
-	keys := make([]interface{}, len(detectors)*2)
-	for i, d := range detectors {
-		keys[i*2] = id
-		keys[i*2+1] = d
+	// TODO(sidac): This kind of type conversion is very useless and wasteful,
+	// we need interfaces around the database models to reduce these kind of
+	// operations.
```

Comment:
TODO: Yeah, rework the interface is very important, the current interface is bad. For example, the strong assumption that the features should be first inserted, and the layers MUST relate to the same features found by the same metadata! Bad.